### ISSUE TITLE
Allow remote imports

### ISSUE THREAD

**Issue by nxy7** [nushell/nushell#9281]
### Related problem

Hello, I've `utils.nu` script that I use across multiple projects I have and I was wondering if remote imports would be something that nushell would support in the future. I don't think that would be very hard to achieve (although it can break LSP) and if maintainers agreed that's a feature they want to support and it's doable (and probably not very hard to do as it would be my first Nushell PR) I could try contributing (once I have a green light and some time to get familiar with codebase). 

### Describe the solution you'd like

Allow syntax like `use https://github.com/nxy7/nushell_utils/utils.nu` that would allow accessing git repos or any text file. Not sure which one is better.

### Describe alternatives you've considered

Git submodules for commonly used scripts.

**Comment by fdncred** [comment]
It seems interesting to me. We already have the `http get` command that can download things, so maybe the guts of that could be used to load a file into memory and "use" it?

**Comment by amtoine** [comment]
we've also started working on a package manager for Nushell, so that might be another alternative :+1: 

> see [`nushell/nupm`: *The Roadmap of the Nushell package manager.*](https://github.com/nushell/nupm/issues/1)

the idea would be to
```
nupm add <pkg>
```
in a project and then be able to 
```
use <pkg>
```
inside said project :smirk:

**Comment by nxy7** [comment]
Package manager would be good if nushell scripts would get more complicated.  Personally I'm not sure they will unless people don't treat it as scripting language and treat it as programming language which might be possible, but there'd be a lot of bias to overcome ^_^
I think both remote imports and package system might fulfill the same purpose, but unless you can start script without running `nump i` first I think remote imports would be much more practical. There's also much less context switching with `use http://someurl` than `ctrl+c; nump add somepkg; openeditor; findFile; use somepkgs` (no background jobs in nushell yet :P).
Either way I'd be excited to get any of those two solutions. I'm wondering what packages would be the most popular for scripting environment 🤔

**Comment by nxy7** [comment]
Hello again, sorry for asking again, but I'd like to have some confirmation from maintainers about this. Is it something that given pull request could be merged into nushell? I have perfect usage for this feature and I'd happily try my luck implementing this, but it would be nice to know that it's something you want to be a part of nushell before I start ^_^

**Comment by fdncred** [comment]
I could see supporting something like `use (http get https://some/url/to/a/script.nu)`.

**Comment by nxy7** [comment]
I think this solution kind of feels like a hack. It's not bad and certainly would be useful, but I was thinking about something better integrated with nushell. If I understand correctly you'd want to allow 'using' output of any command (and `http get` is one of those that would allow remote imports). It's probably the fastest way to make it, but would be harder to build upon later on.
One issue I see with this approach is that remote scripts would not be composable. If you had 'use otherscript.nu' in https://some/url/to/a/script.nu then the whole thing would stop working right?
Imao to make it a worthwhile feature it'd have to be treated same as local imports, so  imports in remote files should work correctly. I don't mind syntax, so if there'd have to be some additional keywords to indicate that it's remote import that's fine (although using some regex for that would make cleaner syntax).

**Comment by fdncred** [comment]
I could see this working too `use https://some/url/to/a/script.nu` but @kubouch is the one who knows `use` best.

**Comment by kubouch** [comment]
I don't think supporting URLs in `use` is a good idea because it would require the parser to send HTTP requests during parsing. It would introduce significant overhead to everything that uses the parser, such as LSP, syntax highlighting, etc. We'd have to handle cases when you don't have the internet or the URL is not available. Seems too complex, slow, and error-prone to me. I think a package manager would be a better solution.

**Comment by nxy7** [comment]
Wouldn't that be the problem only IF particular script uses remote imports? I personally think that SHELL scripting language could benefit from lightweight way of importing stuff from web, but I can also see why one could think it's kind of in the same space as package manager. In my opinion it's not. For complex things package manager is okay, but if you write one time script bringing whole dev environment into it seems not the most ergonomic.
I suspect that package manager would come with some lockfile and something like package.json (and hopefully avoids node_modules situation). So now you need to dedicate whole folder to this onetime script. 
Remote imports with use would be super convenient in this situation (so whole folder turns into just one file with `use http:xxx`. I think there are bigger problems than additional time spent on parsing (as that's the case only if specific file would use this feature). The biggest problem I see is that it'd be really easy to use and for someone to swap the file with malicious code. If both systems would theoretically coexist many people could use remote imports just because it's much easier than packaging code. 
As far as I can see maintainers lean more toward package manager now, if that's somehow as easy to use, then that would be probably the best of both worlds.

**Comment by kubouch** [comment]
An important thing to note is that Nushell [works more like a compiled language](https://www.nushell.sh/book/thinking_in_nu.html#think-of-nushell-as-a-compiled-language) than a traditional shell / dynamic language. I really wouldn't put any network access into the parsing step. As soon as you'd put a file that calls `use` on a URL, it could slow down the parsing massively and require network access for the parsing to even succeed.

Eventually, we plan to ship with `nupm` by default. It's still very early in development, I can imagine there could be an option to let you grab a single file from the internet. Without `nupm`, you could probably code such a tool quite easily (grab a file, put it into a directory that's under NU_LIB_DIRS).

**Comment by nxy7** [comment]
I'm not saying that making something to download file from web is hard, I'm saying that it would be great feature if that would be convenient in shell scripting language. I see that other people don't agree that it's worth having it inside the language, so feel free to close the issue or keep it open in case someone would want to revisit this idea some day.

**Comment by jntrnr** [comment]
Allowing remote imports is something we should avoid. They are a heavy burden on the highlighter, and other kinds of ide support. We should only do the work of parsing during parse time. Anything being that is too much.

**Comment by cyruseuros** [comment]
Unpopular opinion, but personally I am _tired_ of the number of package managers I have. I get that they were a major step forward once upon a time, but now they're like apps - everyone wants you to install one, even though you don't need one.

Personally I love this - the way [deno](https://docs.deno.com/runtime/tutorials/manage_dependencies) does it. That's why it's so great for slapping together cloud functions and whatnot without spinning up a whole project; and if you want to, you can - just create a `deps.ts`. And the whole time you get to forget about caching, modules etc.

Nu is a programming language that is _also a shell_. Shells need to be great for slapping stuff together - if I'm whipping out a package manager, I'm sorry, but I'll go for python, or won't bother with packages.

I understand this is a bit more complicated because of the way nu does parsing, but imagine how amazing this would make nu for devops work (which is where I think it will really shine post 1.0, with all the config files, the process management...)

**Comment by sophiajt** [comment]
@cyruseuros - 

Since I wrote my initial thoughts, I've gone and talked to multiple folks who worked on various packaging systems. The sense I get from them is that the deno way has shortcomings, and the deno folks are continuing to work to overcome those shortcomings. Any time you need to maintain a script, you'll need to do it manually. Even moreso if that single script becomes a project, which we hope people will do for nushell since it's good for more than just the one-off script. They tell me that once you're managing multiples files, it's better to have that management central. It's also better to have a lockfile so that the same project can be reproducibly run in the future.

Something I'm not sure about here is your comment about not wanting another package manager. But the deno way is another package manager. The only difference is you're putting the URL (and likely version) into the import itself. In addition to this, we'd still need to resolve dependencies of those imports and perform version resolution and package management. To my eyes, this feels like one of those things that appears simple on its face, but actually increases the maintenance burden over time on both the developer and the ad-hoc packaging system.

**Comment by cyruseuros** [comment]
@sophiajt I agree with you - It is another package manager, but in terms of UX it basically doesn't exist untill you need it. In that sense you stick to the level of necessary complexity only from the user's perspective.

Re: central management - in deno, if the user needs that kind of power, they have it. Maintaining dependencies and re-exporting them is a common pattern for larger projects (re-exports are already supported really well in nushell)

Re: [lock files](https://docs.deno.com/runtime/manual/basics/modules/integrity_checking) - those are implemented too should you need them (and I would argue most nu projects won't). Although once lock files are enabled - which should probably be done inside deps.nu - the lockfile should automatically be updated when dependencies are added (unlike deno).

It also feel like this functionality is at least as close to the core of the language as an lsp server is, and hence should be part of the core binary.
